stdlib/public/core/AnyHashable.swift

```diff
@@ -107,22 +107,33 @@ internal struct _ConcreteHashableBox<Base: Hashable>: _AnyHashableBox {
 /// A type-erased hashable value.
 ///
 /// The `AnyHashable` type forwards equality comparisons and hashing operations
-/// to an underlying hashable value, hiding its specific underlying type.
+/// to an underlying hashable value, hiding the type of the wrapped value.
 ///
 /// You can store mixed-type keys in dictionaries and other collections that
 /// require `Hashable` conformance by wrapping mixed-type keys in
 /// `AnyHashable` instances:
 ///
 ///     let descriptions: [AnyHashable: Any] = [
-///         AnyHashable("😄"): "emoji",
 ///         AnyHashable(42): "an Int",
 ///         AnyHashable(Int8(43)): "an Int8",
 ///         AnyHashable(Set(["a", "b"])): "a set of strings"
 ///     ]
-///     print(descriptions[AnyHashable(42)]!)      // prints "an Int"
-///     print(descriptions[AnyHashable(45)])       // prints "nil"
+///     print(descriptions[AnyHashable(42)]!)       // prints "an Int"
+///     print(descriptions[AnyHashable(Int8(42))]!) // prints "an Int"
 ///     print(descriptions[AnyHashable(Int8(43))]!) // prints "an Int8"
+///     print(descriptions[AnyHashable(44)])        // prints "nil"
 ///     print(descriptions[AnyHashable(Set(["a", "b"]))]!) // prints "a set of strings"
+///
+/// If the wrapped value is of a type that can be bridged using `as` to a
+/// different but compatible class defined in Foundation, or vice versa, then
+/// the underlying hashable value to which operations are forwarded might not be
+/// of the wrapped type itself but of a transitively bridged type:
```

Comment:
I think it makes sense to spell out exactly  what’s happening here — something akin to “For example, the following three arrays can all bridge to NSArrays of NSNumber, so when boxed in AnyHashable, they compare and hash as if they are equivalent values.”

Reply:
I don't think we want to promise anything about the hash encoding used by AnyHashable. We want to reserve the right to change AnyHashable to add a type discriminator into its hash encoding, in case hash collisions between values of different types ever become a problem. (Adding a type discriminator would mean that converting to/from AnyHashable would *always* change hashes.)

On the other hand, if AnyHashable would guarantee that conversions from/to specific types were hash-preserving, we could use that to speed up certain casts between Set/Dictionary types. So there is a nontrivial trade-off between performance and safety here -- and this tradeoff ought to be carefully considered before we commit to any particular behavior.

I'd replace this paragraph with an explanation that makes fewer concrete promises.

> `AnyHashable` treats types that support as/as?-conversions between them as equivalent, by internally transforming their values into a canonical representation. For instance, `AnyHashable` considers bridged counterparts of the same value (such as a `String` and an `NSString`, or an `Int` and an `NSNumber`) to be equal. Bridging sometimes changes the definition of equality, so that values that were originally distinct may be considered equal when converted to `AnyHashable`:
>
> ```swift
> let string1: NSString = "café"
> let string2: NSString = "cafe\u{301}" // U+301 COMBINING ACUTE ACCENT
> print(string1 == string2) // prints "false"
> print((string1 as AnyHashable) == (string2 as AnyHashable)) // prints "true"
> ```
>
> Note that `AnyHashable` is not guaranteed to preserve the hash encodings of the original values (no matter their type), and the hash encodings it uses may change between standard library releases. Do not assume that AnyHashable generates hash values matching those produced by the original types, even if they appear interchangeable in a particular release.

Reply:
Excellent point @lorentey. I've updated the documentation along those lines. I also split the detailed discussion you mentioned about equality into documentation for `==`.